pymc_experimental/distributions/timeseries.py

```diff
@@ -0,0 +1,214 @@
+                    f"got {type(init_dist)}"
+                )
+            check_dist_not_registered(init_dist)
+            if init_dist.owner.op.ndim_supp > 1:
+                raise ValueError(
+                    "Init distribution must have a scalar or vector support dimension, ",
+                    f"got ndim_supp={init_dist.owner.op.ndim_supp}.",
+                )
+        else:
+            warnings.warn(
+                "Initial distribution not specified, defaulting to "
+                "`Categorical.dist(p=pt.full((k_states, ), 1/k_states), shape=...)`. You can specify an init_dist "
+                "manually to suppress this warning.",
+                UserWarning,
+            )
+            k = P.shape[0]
+            init_dist = pm.Categorical.dist(p=pt.full((k,), 1 / k), shape=batch_size)
+
+        # We can ignore init_dist, as it will be accounted for in the logp term
+        init_dist = ignore_logprob(init_dist)
+
+        return super().dist([P, steps, init_dist], **kwargs)
+
+    @classmethod
+    def rv_op(cls, P, steps, init_dist, size=None):
+        batch_size = size or (1,)
+
+        if init_dist.owner.op.ndim_supp == 0:
+            init_dist_size = (*batch_size,)
+        else:
+            init_dist_size = batch_size
+
+        init_dist = change_dist_size(init_dist, init_dist_size)
+
+        init_dist_ = init_dist.type()
+        P_ = P.type()
+        steps_ = steps.type()
+
+        state_rng = pytensor.shared(np.random.default_rng())
+
+        def transition(previous_state, transition_probs, old_rng):
+            p = transition_probs[previous_state]
+            next_rng, next_state = pm.Categorical.dist(p=p, rng=old_rng).owner.outputs
+            return next_state, {old_rng: next_rng}
+
+        markov_chain, state_updates = pytensor.scan(
+            transition,
+            non_sequences=[P_, state_rng],
+            outputs_info=[init_dist_],
+            n_steps=steps_,
+        )
+
+        (state_next_rng,) = tuple(state_updates.values())
+
+        discrete_mc_ = pt.moveaxis(
+            pt.concatenate([init_dist_[None, ...], markov_chain], axis=0), 0, -1
+        ).squeeze()
+
+        discrete_mc_op = DiscreteMarkovChainRV(
+            inputs=[P_, steps_, init_dist_],
+            outputs=[state_next_rng, discrete_mc_],
+            ndim_supp=1,
+        )
+
+        discrete_mc = discrete_mc_op(P, steps, init_dist)
+        return discrete_mc
+
+
+@_change_dist_size.register(DiscreteMarkovChainRV)
+def change_mc_size(op, dist, new_size, expand=False):
+    if expand:
+        old_size = dist.shape[:-1]
+        new_size = tuple(new_size) + tuple(old_size)
+
+    return DiscreteMarkovChain.rv_op(
+        *dist.owner.inputs[:-1],
+        size=new_size,
+    )
+
+
+@_logprob.register(DiscreteMarkovChainRV)
+def discrete_mc_logp(op, values, P, steps, init_dist, state_rng, **kwargs):
+    n, k = P.shape
```

Comment:
```suggestion
    n, k = P.shape[-2:]
```

Reply:
In the ndim > 2 case every dimension needs to be the same, so it should be changed to something like `pt.all(pt.eq(P.shape, P.shape[0])`

Reply:
Actually if there's the potential for batch dimensions on `P` the whole logp function needs some tweaking, not just the shape check.

Reply:
Not all dims need to be the same. P.shape=[2, 3, 3] would mean 2 independent chains, each with 3 states. Only the last two shapes need to match (be square)

Reply:
When I looked at the logp, I thought it would work out of the box.

Reply:
As currently implemented, the scan isn't written to handle independent chains. I just assumed P would always be a square tensor. I will experiment to see what changes are required, if any.